src/java.base/share/classes/sun/launcher/LauncherHelper.java

```diff
@@ -594,18 +592,26 @@ static void initOutput(PrintStream ps) {
         ostream = ps;
     }
 
-    private static String getMainClassFromJar(JarFile jarFile) throws IOException {
-        String jarname = jarFile.getName();
-        Manifest manifest = jarFile.getManifest();
-        if (manifest == null) {
-            abort(null, "java.launcher.jar.error2", jarname);
-        }
-
+    /**
+     * Expects the presence of a {@code Main-Class} attribute in the jar's manifest
+     * and returns that value. If the attribute is absent then this method calls
+     * {@link #abort(Throwable, String, Object...)}. After ascertaining the presence of
+     * {@code Main-Class}, if {@code JavaFX-Application-Class} attribute is found, then
+     * this method returns the value of {@code JavaFX-Application-Class} after appropriately
+     * {@linkplain FXHelper#setFXLaunchParameters(String, int) configuring} Java FX runtime.
+     *
+     * <p>
+     * Additionally, this method also checks for the presence of {@code Launcher-Agent-Class}
+     * attribute in the manifest. If found, and if {@code java.instrument} module is available,
+     * then the agent will loaded. If there's any exception when trying to load the agent, then
+     * this method will call {@code abort()}.
+     * <p>
+     * {@code Add-Exports} and {@code Add-Opens} are also parsed and processed by this method.
+     *
+     * @return the values of the {@code Main-Class} attribute
+     */
+    private static String parseManifest(String jarname, Manifest manifest) throws IOException {
         Attributes mainAttrs = manifest.getMainAttributes();
-        if (mainAttrs == null) {
-            abort(null, "java.launcher.jar.error3", jarname);
-        }
```

Comment:
Why do we no longer need a null check?

Reply:
The null check is still done, but it's now done before calling this private method. It's done here https://github.com/openjdk/jdk/pull/20929/files/abc3100078c2ab522054d1074d1cdc53c2879388#diff-108a3a3e3c2d108c8c7f19ea498f641413b7c9239ecd2975a6c27d904c2ba226R848